src/gluonts/dataset/loader.py

```diff
@@ -194,99 +157,110 @@ def win32_guard(num_workers: Optional[int]) -> Optional[int]:
-                dataset,
-                transform=transform,
-                shuffle_buffer_length=shuffle_buffer_length,
-            )
-            self.batch_iterator = map(stack_fn, batcher(iterator, batch_size))
-        else:
-            self.batch_iterator = MultiProcessBatcher(
-                dataset,
-                transform=transform,
-                batch_size=batch_size,
-                stack_fn=stack_fn,
-                decode_fn=decode_fn,
+    def __iter__(self):
+        batch_iterator = (
+            map(self.stack_fn, batcher(self.data_iterable, self.batch_size))
+            if not self.num_workers
+            else MultiProcessBatcher(
+                self.data_iterable,
+                batch_size=self.batch_size,
+                stack_fn=self.stack_fn,
+                decode_fn=self.decode_fn,
                 num_workers=self.num_workers,
-                max_queue_size=num_prefetch,
-                shuffle_buffer_length=shuffle_buffer_length,
+                max_queue_size=self.num_prefetch,
             )
+        )
 
-    def __iter__(self):
-        yield from self.batch_iterator
+        return batch_iterator
 
 
-class ValidationDataLoader(DataLoader):
-    def __init__(
-        self,
-        dataset: Dataset,
-        *,
-        transform: Transformation,
-        batch_size: int,
-        stack_fn: Callable,
-        # FIXME: the following aren't used
-        num_workers: Optional[int] = None,
-        num_prefetch: Optional[int] = None,
-        shuffle_buffer_length: Optional[int] = None,
-    ) -> None:
-        self.transformed_dataset = TransformedDataset(
-            dataset,
-            transform,
-            is_train=True,
-        )
-        self.batch_size = batch_size
-        self.stack_fn = stack_fn
+# TODO: the following are for backward compatibility, and could eventually be removed
 
-    def __iter__(self):
-        yield from map(
-            self.stack_fn,
-            batcher(self.transformed_dataset, self.batch_size),
+
+def TrainDataLoader(
+    dataset: Dataset,
+    *,
+    transform: Transformation,
+    batch_size: int,
+    stack_fn: Callable,
+    num_batches_per_epoch: Optional[int] = None,
+    num_workers: Optional[int] = None,
+    num_prefetch: Optional[int] = None,
+    shuffle_buffer_length: Optional[int] = None,
+    decode_fn: Callable = lambda x: x,
+):
+    transformed_dataset = TransformedDataset(
+        cyclic(dataset), transform, is_train=True
+    )
+    data_iterable = (
+        pseudo_shuffled(
+            transformed_dataset, shuffle_buffer_length=shuffle_buffer_length
         )
+        if shuffle_buffer_length is not None
+        else transformed_dataset
+    )
+    data_loader = DataLoader(
+        data_iterable=data_iterable,
+        batch_size=batch_size,
+        stack_fn=stack_fn,
+        num_workers=num_workers,
+        num_prefetch=num_prefetch,
+        decode_fn=decode_fn,
+    )
+    return (
```

Comment:
This looked like a bug at first (I thought you could only iterate once over the `TrainDataLoader`), because it's a bit unexpected that `iterable_slice` is a class. Should we maybe make all classes camel case?

Also one should probably return the `data_loader` instance instead of `iter(data_loader)`?

Reply:
> Should we maybe make all classes camel case?

Yes, why not

> Also one should probably return the data_loader instance instead of iter(data_loader)?

The thing is, that `DataLoader` starts from scratch every time it’s iterated over. While this is fine for validation and prediction, at training time we want it to keep going from where it left instead: with `iter` we’re exactly returning such a “stateful” object.

Reply:
The change from generators to classes was necessary since otherwise these itertools wouldn’t be pickleable, which would make them useless with multiprocessing when “spawn” is used.

Before we were just passing everything down to the processes, which would then construct the generators themselves, but I think this way it’s much more elegant.

Reply:
> > Should we maybe make all classes camel case?
> 
> Yes, why not
> 
> > Also one should probably return the data_loader instance instead of iter(data_loader)?
> 
> The thing is, that `DataLoader` starts from scratch every time it’s iterated over. While this is fine for validation and prediction, at training time we want it to keep going from where it left instead: with `iter` we’re exactly returning such a “stateful” object.
> 
> So essentially, without `iter` you will not go over the entire dataset in case an epoch is too short to cover it, while `iter` allows it.

Mmh, but this is only the case when `num_batches_per_epoch` is set, right? Otherwise you always loop over the entire dataset. If you return `iter(data_loader)` how do you go over the dataset again in the next epoch? 
(I'm probably missing something)

Reply:
> Mmh, but this is only the case when `num_batches_per_epoch` is set, right?

I think currently that has to be set: `num_batches_per_epoch: int = 50`

Reply:
Ok, so then I guess we should not have the `if num_batches_per_epoch is None` case.

Reply:
> Ok, so then I guess we should not have the `if num_batches_per_epoch is None` case.

Sorry: I meant `num_batches_per_epoch: int = 50` in the `Trainer` class. Here, the sequence is infinite if you don't set `num_batches_per_epoch`, in which case the caller should limit the "chunks" it uses on its side (this is what the `Trainer` does)